restapi/src/main/java/io/stargate/web/docsapi/service/query/condition/impl/CombinedCondition.java

```diff
@@ -64,11 +64,20 @@ protected void validate() {
   /** {@inheritDoc} */
   @Override
   public boolean test(Row row) {
-    Object dbValue = getValue(row, isNumericBooleans());
+    Boolean dbValueBoolean = getBoolean(row, isNumericBooleans());
+    Double dbValueDouble = getDouble(row);
+    String dbValueString = getString(row);
 
     CombinedFilterOperation<V> filterOperation = getFilterOperation();
     V queryValue = getQueryValue();
 
-    return filterOperation.test(queryValue, dbValue);
+    // compare against the non-null values, fallback to text compare even if null
+    if (null != dbValueBoolean) {
```

Comment:
It seems off to me to do a `Row#isNull()` then do another check for `!= null`  (then again in the filter operation).

Reply:
It might be better to do something like:

```java
public interface GenericFilterOperation<FV> extends BaseFilterOperation {
  boolean test(FV filterValue, Row row, boolean isNumericBoolean);
}
```

```java
public abstract class InFilterOperation implements GenericFilterOperation<List<?>> {

  @Override
  public boolean test(List<?> filterValue, Row row, boolean isNumericBoolean) {
    // if null, check any is null, avoid .contains(null) as some impl could throw NPE
    if (row.isNull(QueryConstants.BOOLEAN_VALUE_COLUMN_NAME)) {
      Boolean b = row.getBoolean(QueryConstants.BOOLEAN_VALUE_COLUMN_NAME);
      // Do stuff and deal with numeric bool
    } else if (row.isNull(QueryConstants.DOUBLE_VALUE_COLUMN_NAME)) {
      Double d = row.getDouble(QueryConstants.DOUBLE_VALUE_COLUMN_NAME);
      // Do stuff
    } else if (row.isNull(QueryConstants.STRING_VALUE_COLUMN_NAME)) {
      String s = row.getString(QueryConstants.BOOLEAN_VALUE_COLUMN_NAME);
      // Do stuff
    }
    return filterValue.stream().anyMatch(Objects::isNull);
  }
}
```

This would also allow for the removal of the row specific code from `BaseCondition`  (which is sort of out of place) and it could be moved into each specific condition type e.g.

```java
public abstract class NumberCondition implements BaseCondition {

  @Override
  public boolean test(Row row) {
    Double dbValue = row.isNull(QueryConstants.DOUBLE_VALUE_COLUMN_NAME) ? null :
        row.getDouble(QueryConstants.DOUBLE_VALUE_COLUMN_NAME);
    return getFilterOperation().test(getQueryValue(), dbValue);
  }
}

Reply:
Could we also get rid of the numeric boolean settings by adding a utility function like?

```java
public static getBoolean(Row row) {
  int index = row.firstIndexOf(QueryConstants.BOOLEAN_VALUE_COLUMN_NAME);
  if (!row.isNull(index)) {
    ColumnType type = row.columns().get(index).type().rawType();
    return type == Type.Boolean ? row.getBoolean(index) : (row.getByte(index) != 0);
  } else {
    return null;
  }
}
```

Reply:
There are few suggestions here, so let me comment back on all of them:

* the whole intention was that the `FilterOperation` is not depending on the Row or any other data store related structure.. So I would like to keep that, FilterOperation should not know anything about the data store.. One benefit would be that we could re-use filter operations in other peaces of code, making a unified, well tested filtering available.
* having row specific code in the `BaseCondition` was intentional in order to keep it all in one single class and re-usable by all implementing ones   
* idea about the numeric boolean is not bad at all, @dimas-b @EricBorczuk  what do you think?

Reply:
> the whole intention was that the FilterOperation is not depending on the Row or any other data store related structure.. So I would like to keep that, FilterOperation should not know anything about the data store.. One benefit would be that we could re-use filter operations in other peaces of code, making a unified, well tested filtering available.

A better way to achieve that would be to wrap `Row` in a `Value`-like interface:

```java
interface Value {
  boolean isBoolean();
  Boolean getBoolean();
  boolean isDouble();
  Double getDouble();
  // etc.
}
```

Then:

```java
public interface GenericFilterOperation<FV> extends BaseFilterOperation {
  boolean test(FV filterValue, Value value);
}
```

Reply:
> having row specific code in the BaseCondition was intentional in order to keep it all in one single class and re-usable by all implementing ones

Are we planning to have many more of these? It seems off to have concrete details in the base interface.

Reply:
All this said, I'm fine with moving on. :)